airflow/contrib/sensors/emr_job_flow_sensor.py

```diff
@@ -16,48 +16,96 @@
 # KIND, either express or implied.  See the License for the
 # specific language governing permissions and limitations
 # under the License.
+
 from airflow.contrib.hooks.emr_hook import EmrHook
 from airflow.contrib.sensors.emr_base_sensor import EmrBaseSensor
 from airflow.utils.decorators import apply_defaults
 
 
 class EmrJobFlowSensor(EmrBaseSensor):
     """
-    Asks for the state of the JobFlow until it reaches a terminal state.
+    Asks for the state of the EMR JobFlow (Cluster) until it reaches
+    any of the target states.
     If it fails the sensor errors, failing the task.
 
+    With the default target states, sensor waits cluster to be terminated.
+    When target_states is set to ['RUNNING', 'WAITING'] sensor waits
+    until job flow to be ready (after 'STARTING' and 'BOOTSTRAPPING' states)
+
     :param job_flow_id: job_flow_id to check the state of
     :type job_flow_id: str
+
+    :param target_states: the target states, sensor waits until
+        job flow reaches any of these states
+    :type target_states: list[str]
+
+    :param failed_states: the failure states, sensor fails when
+        job flow reaches any of these states
+    :type failed_states: list[str]
     """
 
-    NON_TERMINAL_STATES = ['STARTING', 'BOOTSTRAPPING', 'RUNNING',
-                           'WAITING', 'TERMINATING']
-    FAILED_STATE = ['TERMINATED_WITH_ERRORS']
-    template_fields = ['job_flow_id']
+    template_fields = ['job_flow_id', 'target_states', 'failed_states']
     template_ext = ()
 
     @apply_defaults
     def __init__(self,
                  job_flow_id,
+                 target_states=None,
+                 failed_states=None,
                  *args,
                  **kwargs):
         super().__init__(*args, **kwargs)
         self.job_flow_id = job_flow_id
+        if target_states is None:
+            target_states = ['TERMINATED']
```

Comment:
What about oneliner?
```suggestion
            target_states = target_states or ['TERMINATED']
```

And shouldn't we extend the list to `['STARTING', 'BOOTSTRAPPING', 'RUNNING', 'WAITING', 'TERMINATING']` to preserve backward compatibility?

Reply:
Actually I have reversed the state logic, previously it was checking whether the state is in the `NON_TERMINAL_STATES` or not, if it is in the list, it was returning `False`. Now it is checking whether the state is in the `self.target_states` or not, if it is in it returns `True`. I could say `self.non_target_states` but is does not seem well.
I am sure that it is bacward compatible with this list, and if it was not, the unit test would fail. (I have also updated them to simplify, there were 4 classes almost same, I changed them with 1 common class).

oneliner is good idea, I will do it.